### PR TITLE
syslog: fix log confusion when multi task writing together

### PR DESCRIPTION
**Pull request by Donny9**
## Summary
Let's to do "syslog_write_t sc_write" function to fix log confusion when multi task writing together for syslog and ramlog.
## Impact
normal output log for multi task environment.
## Testing
daily test.

### CODE REVIEW

**Comment by fjpanag** [comment]
Hi @Donny9 

I had a look at the changes, but I don't think that this will fix the issue.  
The problem is not the thread safety of the drivers, rather the multiple calls to `lib_sprintf()`.  
See #3599.

Adding a `sc_write()` to all channels seems a good way forward.
If `syslog_write_t` is modified to also accept a `force` flag, we can eliminate the `syslog_putc_t` too.

But I think a semaphore at this place will not solve the actual issue. Depending on the solution to the initial problem, it may not be needed at all.

Do you have any plans for further development on this?  
(I have also spent quite some time thinking about this and experimenting, but I haven't reached a "good" final implementation...)

**Comment by xiaoxiang781216** [comment]
> Hi @Donny9
> 
> I had a look at the changes, but I don't think that this will fix the issue.
> The problem is not the thread safety of the drivers, rather the multiple calls to `lib_sprintf()`.
> See #3599.

lib_sprintf format result to lib_syslogstream_s, two cases here:

1. CONFIG_SYSLOG_BUFFER equals y, syslogstream_flush output the log through syslog_write
2. CONFIG_SYSLOG_BUFFER equals n, syslogstream_addchar output the log through syslog_putc

This patch only address the mess issue in case 1. The lock ensure that the caller of syslog_write don't interlace with each other.

> 
> Adding a `sc_write()` to all channels seems a good way forward.
> If `syslog_write_t` is modified to also accept a `force` flag, we can eliminate the `syslog_putc_t` too.
>

Yes, but sc_putc/sc_force_putc is more simpler to implement than sc_write.
 
> But I think a semaphore at this place will not solve the actual issue. Depending on the solution to the initial problem, it may not be needed at all.
> 
> Do you have any plans for further development on this?
> (I have also spent quite some time thinking about this and experimenting, but I haven't reached a "good" final implementation...)

Case 1 should be addressed by:

1. syslog driver self(e.g. ramlog, devlog), if it provide the multithread safe sc_write
2. syslog_default_write with this path

But, it's impossible to fix case 2.

**Comment by fjpanag** [comment]
@xiaoxiang781216 
> This patch only address the mess issue in case 1. The lock ensure that the caller of syslog_write don't interlace with each other.

Don't forget that also buffering needs fixing. In case of overflow it will split the message in multiple `syslog_write` calls, which again leaves room for this bug to appear....

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> 
> > This patch only address the mess issue in case 1. The lock ensure that the caller of syslog_write don't interlace with each other.
> 
> Don't forget that also buffering needs fixing. In case of overflow it will split the message in multiple `syslog_write` calls, which again leaves room for this bug to appear....

Yes, it can't fix too, we need reserve at least one line buffer.

**Comment by xiaoxiang781216** [comment]
@fjpanag do you think this patch fix some case of mess after discussion?

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by fjpanag** [comment]
@xiaoxiang781216 Sorry for the late response, I just had the time to check on the code.

Yes, I believe it is better.  
At least some cases are now protected.

File: drivers/syslog/ramlog.c [hunk 1]
```diff
@@ -283,12 +286,89 @@ static int ramlog_addchar(FAR struct ramlog_dev_s *priv, char ch)
...
+   * number of bytes that were actually written.  Otherwise, callers
+   * probably retry, causing same error condition again.
+   */
+
+  nxsem_post(&priv->rl_exclsem);
+  return len;
+}
+
 /****************************************************************************
  * Name: ramlog_read
```
Comment (fjpanag) [inline drivers/syslog/ramlog.c:367]: Just a minor nit. The comment was not updated to reflect the new function's name.

**Pull request merged**